### PR TITLE
opt,sql: support streaming set ops

### LINKED ISSUE #40797: opt: support streaming set ops
INTERSECT and EXCEPT are currently executed by HashJoiners (with special "join" modes). The distsql planner supports using MergeJoiners as well, but the optimizer is not aware of this and does not set up orderings.

This issue tracks adding this intelligence into the optimizer and populating the `unionNode` with the ordering information.

### PR DESCRIPTION
**Pull request by rytaft**
This commit adds support for planning streaming set operations in
the optimizer. Streaming set operations consist of a `UNION`, `INTERSECT`,
or `EXCEPT` (including `ALL` variants) in which both inputs are ordered
on the desired output ordering, and the set operation merges the
inputs to maintain the ordering. The execution engine already
supported this type of streaming set operation, so the purpose of
this commit was to teach the optimizer about this capability and hook
it up to the execution engine.

This is also a prerequisite for #56201.

Fixes #40797
Informs #56201

Release note (performance improvement): Set operations (`UNION`,
`UNION ALL`, `INTERSECT`, `INTERSECT ALL`, `EXCEPT` and `EXCEPT ALL`) can
now maintain ordering if both inputs are ordered on the desired
output ordering. This can eliminate unnecessary sort operations
and improve performance.

### CODE REVIEW

**Review by mgartner — APPROVED** [review]
:lgtm: 

Very cool! Talk about a high-leverage commit; a few new lines of code and some great query plan improvements.

> Reviewed 25 of 25 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYfMiXgDj7zsIeoK4R2:b-6nh2cp)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde, @rytaft, and @yuzefovich)

___
*[pkg/sql/opt/xform/testdata/external/trading, line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYfL6kTAY3RRAKgh7fo:-MYfL6kTAY3RRAKgh7fp:b-hjzsaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8553b79c79e91ee1e129d98a965bb921a142362/pkg/sql/opt/xform/testdata/external/trading#L606)):*
> ```
>  │    │    ├── key: ()
>  │    │    ├── fd: ()-->(8,16)
>  │    │    ├── sort
> ```

It's great to see these `sort`s just disappear! :tada: 
___
*[pkg/sql/opt/xform/testdata/rules/limit, line 905 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYfMGB7Ao2EfqdDwpWG:-MYfMGB7Ao2EfqdDwpWH:b-1j2pqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d36b1e8d2af29c2a1d64c0742fff6e69834d0f/pkg/sql/opt/xform/testdata/rules/limit#L905)):*
> ```
>  │    │    │    ├── key: ()
>  │    │    │    ├── fd: ()-->(14,17)
>  │    │    │    └── limit hint: 1.00
> ```

I'm not sure I've seen this before: a limit hint and a hard limit. In this case would the limit hint have any effect on what the execution engine does?

**Review by yuzefovich — COMMENTED** [review]
I agree, very cool stuff!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYfbhTb-P9F2rrt9ZSs:b1hw866)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/testdata/union, line 260 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYfPZBFC4HLkk_lrWue:-MYfPZBFC4HLkk_lrWuf:bsey29t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d36b1e8d2af29c2a1d64c0742fff6e69834d0f/pkg/sql/opt/exec/execbuilder/testdata/union#L260)):*
> ```
> Diagram: https://cockroachdb.github.io/distsqlplan/decode.html#eJykkc9rq0AQx-_vrxjm9B5vQ9TePCVpLAhWU01LS_GwukMQjGt3DSjB_724tqRCU5r2JDPfHx-HPaJ-KdFF73ETLP0Q_q79ZJvcBQwevHgVJd4_SLzAu94CZ5DBTRzdAs_gPvSjEJZB8K62sEwGSzd833xtB1G89mJYPQFHhpUUFPI9aXSf0caUYa1kTlpLNayOxuCLFl2LYVHVh2ZYpwxzqQjdIzZFUxK6uOVZSTFxQWpuIUNBDS9KU9t2i1oVe646ZJjUvNIuzDDtGcpDc2rUDd8RunbPfka1p1SeXUJ1zlJPMKkEKRJTzML-j2n_ya-FcibruTNxn6Nbl9wck65lpembzSlDEjsaT9HyoHLaKJkbzDhGJmcWgnQzqs44-JWRzKN8DNu_CTtfhq8mYatP-z-vAQAA___WJ_y5
> 
> # TODO(rytaft): Should we be concerned that the synchronizers in the below
> ```

I looked into this for a bit, it is not concerning.

I think what happens is that first, when adding a distinct stage, we see that there is only one stream coming out of it, so we set `mergeOrdering` to empty (in `PhysicalPlan.SetMergeOrdering`); then, later when planning the merge join processor we will create an input synchronizer with only one incoming stream and we choose "unordered" because it is supposed to be more efficient (I think that was the case in the row-based flows). Still, since there is only one stream, it doesn't matter.

It is a bit annoying to see that we have an input synchronizer at all when there is only one stream, but this behavior has been present for a while, so not a concern of this PR. Feel free to leave a TODO to look into it (also feel free to put my name on it).
___
*[pkg/sql/opt/xform/testdata/rules/limit, line 905 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYfMGB7Ao2EfqdDwpWG:-MYfVaIDAXawMoO9gI_y:b-hw9kyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d36b1e8d2af29c2a1d64c0742fff6e69834d0f/pkg/sql/opt/xform/testdata/rules/limit#L905)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm not sure I've seen this before: a limit hint and a hard limit. In this case would the limit hint have any effect on what the execution engine does?
</blockquote></details>

The execution engine prefers the hard limit if set and ignores the soft limit entirely.

**Review by rytaft — COMMENTED** [review]
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYfmNoW7s9x3ZAWFpOE:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/opt/exec/execbuilder/testdata/union, line 260 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYfPZBFC4HLkk_lrWue:-MYflZqA06ELsD2f576j:bb7ldmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d36b1e8d2af29c2a1d64c0742fff6e69834d0f/pkg/sql/opt/exec/execbuilder/testdata/union#L260)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I looked into this for a bit, it is not concerning.

I think what happens is that first, when adding a distinct stage, we see that there is only one stream coming out of it, so we set `mergeOrdering` to empty (in `PhysicalPlan.SetMergeOrdering`); then, later when planning the merge join processor we will create an input synchronizer with only one incoming stream and we choose "unordered" because it is supposed to be more efficient (I think that was the case in the row-based flows). Still, since there is only one stream, it doesn't matter.

It is a bit annoying to see that we have an input synchronizer at all when there is only one stream, but this behavior has been present for a while, so not a concern of this PR. Feel free to leave a TODO to look into it (also feel free to put my name on it).
</blockquote></details>

Thanks for looking into this! Added a TODO with your name on it.
___
*[pkg/sql/opt/xform/testdata/external/trading, line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYfL6kTAY3RRAKgh7fo:-MYflhkrENKoiDh9U0i9:b-ve3yvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8553b79c79e91ee1e129d98a965bb921a142362/pkg/sql/opt/xform/testdata/external/trading#L606)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It's great to see these `sort`s just disappear! :tada: 
</blockquote></details>

:smile:

**Review by RaduBerinde — COMMENTED** [review]
Nice work!

What happens when we have an ordering only on a subset of cols? I believe the merge joiner can only work when all columns are equality columns, no? So then aren't we assuming that the merge joiner will maintain ordering in all these cases when it might not?

For example:
```
create table ab (a int, b int, index (a) storing (b));
create table de (d int, e int, index (d) storing (e));
explain (distsql) select a,b from ab intersect select d, e from de order by a;
```
We get this [plan](https://cockroachdb.github.io/distsqlplan/decode.html#eJyckt-LnDAQx9_7V4R5uoOUXX_cS6DgtWupYNer-tBSZIlmuidYY5MIV5b934um5U65s7c-Ziaf-X4megL9qwEG4de7-Dbak6tdlOXZl_iaZGEcfsgJp6QkH9PkM-ElifZ5mGZD-W9XUIK2K5Ak6S5MyftvhAOFVgrc85-ogX0HBwoKnZIVai3VUDqNFyLxAGxLoW673gzlgkIlFQI7galNg8Ag52WDKXKBarMFCgINr5txrMBA4EEcavEAFLKOt5qRt0Ah6Q0jgUMDF4ozBdmbx_na8CMCc850nYMzdeBlwMsDX-XgXuKwq7Wp28ps3KmAjaCQKIEKxZD5YqC3KtBbH-i_GPiY07fSTprEFAP5vyvPWH_i-j5Dk3Qbf6qd_-6QPfmDb-MYKDT4w1yNG12_U_Xx_t_hdR_w5pL3TFF3stU4X_PZydthNxRHtG-lZa8qvFOyGmPsMRm5sSBQG9v17CFqbWsQfAo7i7C7DLuLsD-BnTnsXQC7c9hfhG9m2sX5zZ8AAAD__zkyjJM=) which uses a HashSetOp. @yuzefovich does a HashSetOp guarantee anything about the output ordering? My guess is no, if it spills to disk (which is why we don't pass orderings through HashJoin).

Another thing - we still need some more work in this area that should be captured in an issue and/or TODO: we need exploration rules that generate unions with interesting orderings, similar to streaming groupby (currently we can only take advantage of the faster execution if there is an ORDER BY requirement); and the coster needs to make the distinction between the hash and the merge execution.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYfsOLBDNqEHrhtQB4a:bcuev2p)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @rytaft, and @yuzefovich)

___
*[pkg/sql/opt/exec/factory.opt, line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYftckk2F_b8RgBI09a:-MYftckk2F_b8RgBI09b:b7o9y0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d36b1e8d2af29c2a1d64c0742fff6e69834d0f/pkg/sql/opt/exec/factory.opt#L169)):*
> ```
> # SetOp performs a UNION / INTERSECT / EXCEPT operation (either the ALL or the
> # DISTINCT version). The left and right nodes must have the same number of
> # columns. reqOrdering specifies the required output ordering, and if not empty,
> ```

This needs more information, mentioning that when reqOrdering is set we do a streaming operation (and @yuzefovich should sign off on it).

**Review by yuzefovich — COMMENTED** [review]
> What happens when we have an ordering only on a subset of cols? I believe the merge joiner can only work when all columns are equality columns, no?

Yes, that is very good point - the hash join will be used in this case, and it doesn't take into account any partial ordering.

> So then aren't we assuming that the merge joiner will maintain ordering in all these cases when it might not?

The merge joiner does maintain ordering for set-op join types (I believe the only exception is FULL OUTER where the ordering is ambiguous). This will also hold if the spilling to disk occurs.

> We get this plan which uses a HashSetOp. @yuzefovich does a HashSetOp guarantee anything about the output ordering? My guess is no, if it spills to disk (which is why we don't pass orderings through HashJoin).

The hash joiner maintains the ordering of rows for set-operation joins with the only exception of the vectorized hash joiner when it spills to disk (as Radu correctly guessed). In order maintain the partial ordering we'll need to plan an external sort (similar to what we did for the external hash aggregator).

Alternatively, we could only support the ordering propagation for set-ops if the merge joiner is used.

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYfvTG96P3f9QSeOKf6:b-dsvkbk)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

___
*[pkg/sql/opt/exec/factory.opt, line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYftckk2F_b8RgBI09a:-MYfvE5l7NkPmuDcEp5s:b-kfg2yj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d36b1e8d2af29c2a1d64c0742fff6e69834d0f/pkg/sql/opt/exec/factory.opt#L169)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This needs more information, mentioning that when reqOrdering is set we do a streaming operation (and @yuzefovich should sign off on it).
</blockquote></details>

What do you mean a "streaming operation" - that the merge join is used?

**Review by RaduBerinde — APPROVED** [review]
> > So then aren't we assuming that the merge joiner will maintain ordering in all these cases when it might not?
> The merge joiner does maintain ordering for set-op join types (I believe the only exception is FULL OUTER where the ordering is ambiguous). This will also hold if the spilling to disk occurs.

I misspoke, I should have said "the execution operator" (we act as if we're planning a merge joiner from the optimizer, but we're actually planning a hash operation).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYg6Fqb1ThM30l3e5r7:bgotqn)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[pkg/sql/opt/exec/factory.opt, line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYftckk2F_b8RgBI09a:-MYg6BnJC2jCNJBPotGz:bb77u1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d36b1e8d2af29c2a1d64c0742fff6e69834d0f/pkg/sql/opt/exec/factory.opt#L169)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

What do you mean a "streaming operation" - that the merge join is used?
</blockquote></details>

Yeah

**Review by yuzefovich — CHANGES_REQUESTED** [review]
Wait, Radu's example does show a case where we might have a problem - in case when we have a partial ordering, the HashSetOp is planned, and we need to make sure to have the desired output ordering. when we spill to disk in the vectorized engine. We need to do something very similar to what https://github.com/cockroachdb/cockroach/pull/63372/commits/ef1323208db40946714b95ee8e4009b1bc9586a0 did.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYg9qbuAnCgROesnH1g:bpmvxb2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

___
*[pkg/sql/opt/exec/factory.opt, line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYftckk2F_b8RgBI09a:-MYg7tMf-QrAERZT3BKB:bmdfifq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d36b1e8d2af29c2a1d64c0742fff6e69834d0f/pkg/sql/opt/exec/factory.opt#L169)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yeah
</blockquote></details>

Ack. Radu's suggestion sounds good to me.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

___
*[pkg/sql/opt/exec/factory.opt, line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYftckk2F_b8RgBI09a:-MYgAd_2AfDX3vjurSvU:b-yqgsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d36b1e8d2af29c2a1d64c0742fff6e69834d0f/pkg/sql/opt/exec/factory.opt#L169)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Ack. Radu's suggestion sounds good to me.
</blockquote></details>

I guess the operation will be streaming only if `ReqOrdering` contains all columns produced by this `SetOp`.

**Review by RaduBerinde — APPROVED** [review]
I don't think we should add the sort in the execution layer. We did that in the interest of fixing a correctness problem quickly and in backportable fashion. The optimizer needs to know exactly what it is planning and cost accordingly. It should be able to plan segmented sorts in order to get the streaming variant. We should have two exec.Factory, one  HashSetOp and one StreamingSetOp (or MergeSetOp), and the semantics of each should be clearly defined.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYh5kk26bYqUJ5aRYJP:b-hx62gj)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

**Review by yuzefovich — CHANGES_REQUESTED** [review]
We might be saying the same thing, but I'm a bit confused about your "approval" stance. Currently, the following logic test (extending your example a bit more) will fail occasionally:
```
# LogicTest: fakedist-disk

statement ok
create table ab (a int, b int, index (a) storing (b));
create table de (d int, e int, index (d) storing (e));
insert into ab values (1, 1), (2, 2), (3, 3), (4, 4), (5, 5);
insert into de values (1, 1), (3, 3), (5, 5), (7, 7);

query II
select a, b from ab intersect select d, e from de order by a
----
1 1
3 3
5 5
```
because we use the vectorized hash joiner, and when spilled to disk it doesn't maintain the ordering as the optimizer expects.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYjvuio7fPfQAzJea6J:b9rhifh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

**Review by RaduBerinde — COMMENTED** [review]
Approval was not intended, I should have changed the default button in Reviewable. The problem needs to be addressed before merging.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYjxAMS5w3jC_UXWCzC:b-iksrl2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

**Review by rytaft — COMMENTED** [review]
TFTRs! Great catch, @raduberinde, and thanks for the example test case, @yuzefovich! I've updated the logic to ensure that we always order on all columns if the optimizer wants to plan a MergeSetOp.

> Another thing - we still need some more work in this area that should be captured in an issue and/or TODO: we need exploration rules that generate unions with interesting orderings, similar to streaming groupby (currently we can only take advantage of the faster execution if there is an ORDER BY requirement); and the coster needs to make the distinction between the hash and the merge execution.

Opened #64062

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYtcDBP8EhIEBN5gOrr:b9fo9l3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @yuzefovich)

___
*[pkg/sql/opt/exec/factory.opt, line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYftckk2F_b8RgBI09a:-MYoB67kDpMijqWD6fJf:b-a2ov03) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d36b1e8d2af29c2a1d64c0742fff6e69834d0f/pkg/sql/opt/exec/factory.opt#L169)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I guess the operation will be streaming only if `ReqOrdering` contains all columns produced by this `SetOp`.
</blockquote></details>

Done. Let me know if this new text is accurate.

**Review by yuzefovich — APPROVED** [review]
Nice! The physical planner changes LGTM (I haven't looked at the `opt` changes).

> Reviewed 4 of 9 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYvOJB71SugCKQaNOTm:bo88glz)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[pkg/sql/opt/exec/factory.opt, line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-MYftckk2F_b8RgBI09a:-MYvLQ8K7iDK5wa0WLic:b-vs1prh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d36b1e8d2af29c2a1d64c0742fff6e69834d0f/pkg/sql/opt/exec/factory.opt#L169)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Done. Let me know if this new text is accurate.
</blockquote></details>

Looks good!

**Review by RaduBerinde — COMMENTED** [review]
Very nice! :lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYvZhfIDGZj0tFb94mW:bup7q37)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

**Review by rytaft — COMMENTED** [review]
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63805#-:-MYvmo_i29tCk_qbvH4w:b-tlx7ko)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

**Pull request merged**
